lib/Relation.php

```diff
@@ -0,0 +1,262 @@
+     * mapping of column names  where(["name"=>"Philip", "publisher"=>"Random House"]) finds the first row of WHERE name=Philip AND publisher=Random House
+     * raw WHERE statement      where(['name = (?) and publisher <> (?)', 'Bill Clinton', 'Random House'])
+     *
+     * @param int|string|array<string> $needle
+     *
+     * @return Model|null The single row that matches query. If no rows match, returns null
+     */
+    public function where(int|string|array $needle): Model|null
+    {
+        $this->limit(1);
+
+        if (is_array($needle)) {
+            $this->options['conditions'] = $needle;
+            $this->options['mapped_names'] = $this->alias_attribute;
+            $list = $this->table->find($this->options);
+        } else {
+            unset($this->options['mapped_names']);
+            $list = $this->find_by_pk($needle);
+        }
+
+        if (null == $list) {
+            return null;
+        }
+
+        return $list[0];
+    }
+
+    /**
+     * needle is one of:
+     *
+     * empty array              all()       returns all rows in the database
+     * array of primary keys    all([1, 3]) WHERE author_id in (1, 3)
+     * mapping of column names  all(["name"=>"Philip", "publisher"=>"Random House"]) WHERE name=Philip AND publisher=Random House
+     * raw WHERE statement      all(['name = (?) and publisher <> (?)', 'Bill Clinton', 'Random House'])
+     *
+     * @param array<number|mixed|string> $needle An array containing values for the pk
+     *
+     * @return array<Model> All the rows that matches query. If no rows match, returns []
+     */
+    public function all(array $needle = []): array
+    {
+        $list = [];
+
+        if (array_is_list($needle) && count($needle) > 0) {
+            unset($this->options['mapped_names']);
+            try {
+                return $this->find_by_pk($needle);
+            } catch (RecordNotFound $mustBeRawWhereStatement) {
+            }
+        }
+
+        // Only for backwards compatibility with version 1 API
+        $isUsingOriginalFind = false;
+        foreach (Model::$VALID_OPTIONS as $key) {
+            if (array_key_exists($key, $needle)) {
+                $this->options[$key] = $needle[$key];
+                $isUsingOriginalFind = true;
+            }
+        }
+
+        if (!$isUsingOriginalFind) {
+            $this->options['conditions'] = $needle;
+        }
+        $this->options['mapped_names'] = $this->alias_attribute;
+
+        return $this->table->find($this->options);
+    }
+
+    /**
+     * Finder method which will find by a single or array of primary keys for this model.
+     *
+     * @see find
+     *
+     * @param mixed $values An array containing values for the pk
+     *
+     * @throws RecordNotFound if a record could not be found
+     *
+     * @return array<Model>
+     */
+    private function find_by_pk(mixed $values): array
+    {
+        if ($this->table->cache_individual_model ?? false) {
```

Comment:
We don't have to do it now, but at some point we should decide if this should delegate to `Model`, or if `Model` should delegate to this (DRY).

Reply:
Also, there doesn't seem to be any `find_by_pk` method in the Ruby API. I think is something that would be handled by the magic "find_by..." pattern.

Reply:
Removed Model::find_by_pk() by moving implementation of Model::find() to Relation